### ISSUE TITLE
VeraCrypt freezing on dismount

### ISSUE THREAD

**Issue by griesi007** [veracrypt/VeraCrypt#208]
Regularly I am facing an issue where VeraCrypt is freezing on dismount of an file container together with another random application. This happens very randomly (cannot reproduce it willful). This is how it behaves.
- e.g. Notepadd++ is open (this can be any application. It is not Notepad++ specific)
- A file container is mounted with VeraCrypt
- A file from the mounted container is opened with another application (e.g. Excel)
- File is closed. Excel is closed.
- File container is dismounted in VeraCrypt. VerayCrypt progress spinner is spinning forever (freezing).
- All other applications are closed. Notepadd++ which is also frozen is terminated via task manager
- VeraCrypt now continues to dismount  

Environment:
Windows 10 Pro (64-bit): Version 1703 / Buiold 15063.608
VeraCrypt 1.21 (64-bit)

**Comment by Lyoko-Jeremie** [comment]
I met the same problem last week, but it only appeared twice.
i try to reboot, it freezen untill all process stop or i press power button 10sec to forced shutdown.

**Comment by griesi007** [comment]
For me this happens very regularly (almost daily). I noticed this now for a longer period of time. I can´t say when it started. With TrueCrypt I never had that problem

**Comment by Lyoko-Jeremie** [comment]
v1.18a seems like not this question for me.
this issue happend since i update to v1.21 .

may be you can have a try?

**Comment by CriusMacOldenly** [comment]
This problem happened for me a long time ago -- it is not v1.21 specific.
Usually I have Firefox as the second top most application, with veracrypt being the top most application. This causes firefox to no longer respond to the mouse.
Just yesterday, I happened to have EVE Online as the second top most application. EVE Online stopped responding to the mouse. The animations were still working though.
Then I had notepad as the second top most application, and it stopped responding when I tried to dismount veracrypt drive.
I have had this problem with older versions of Veracrypt for at least the last 6 months, probably more. It's as if veracrypt is getting confused, and thinking the second top most application is the window it is waiting on for closing. It doesn't matter what the second top most application is -- it stops being able to receive input, athough it is stable functioning, and when this now frozen application is killed via task manager, veracrypt only then finishes mounting or dismounting its drive.
I believe it's a bug with an older version of veracrypt that keeps getting compiled into the newer versions of veracrypt.

**Comment by jgoemaat** [comment]
I also have this happening with V 1.21, but it only is happening when dismounting a device (full partition), not a regular file container (those dismount fine).

**Comment by griesi007** [comment]
@CriusMacOldenly: For me it happens exactly the same way you described it

**Comment by NumDeP** [comment]
Similar problem. 

It happened after I attempted to close one of three regular file containers. If I remember clearly, I was pasting large files into one of the other two containers and as soon as I dismounted, the browser too became unresponsive and only began to became functional again as VeraCrypt became functional as well, which I can tell you was after very long time.

In the end everything did work alright but again it was a long wait.

**Comment by MrWolf-to** [comment]
I'm having the same problem. Sometimes (fortunately, not always), VeraCrypt freezes on dismounting (completely or for a long time) and any other program hangs (usually Firefox). (Windows 7 64 bit)
As far as I'm concerned, this issue has been present in the latest releases, not in the older ones.

**Comment by nickelr** [comment]
I have the same problem. My VeraCrypt freezes and i cant use my mouse on a random program. The random program works but i can't use it anymore and if i close the random program i can use Veracrypt again.

**Comment by Uli-Armbruster** [comment]
Same probleme here: Windows 10 Pro Creator Fall Update and previous version

**Comment by Kakifrucht** [comment]
Recently updated to the newest version 1.21, and only had the issue rarely before, since then 100% freezes on dismount, and sometimes also on mount.

**Comment by GarwelGarwel** [comment]
I'm not sure about another app, but VeraCrypt has been freezing on dismount more and more often recently. I have to end the task via Task Manager. Windows 10 Home Build 1703, x64, VeraCrypt 1.21.

**Comment by Kakifrucht** [comment]
For me the issues have randomly stopped..
Don't know if a Windows update fixed it. Before I always had to shut down the process via Task Manager.

**Comment by CriusMacOldenly** [comment]
I haven't seen a veracrypt freeze in a while. I don't know how or when. I didn't write down the date it stopped freezing.

**Comment by fungic** [comment]
Also reporting same issue - With Creators update and 1.21. 
Oddly enough if I close down firefox it dismounts no problem

**Comment by JoanneDesign** [comment]
I had the same problem.  I checked their help on their dialog box and that fixed it.  You just need to open VeraCrypt (VC), go to "Settings -> Preferences" and check the option "Don't show wait message dialog when performing operations".  I give VC time to open (like a minute) and time to dismount (the same) and all (so far) has been behaving fine.  Let me know if that works for you too.

**Comment by CriusMacOldenly** [comment]
I remember checking that option because the freezing was so annoying, but it made no difference. Veracrypt was still freezing other programs constantly. I've examined "Don't show wait message dialog when performing operations", and it is still checked. While the freezing has gone away, I know that option isn't what fixed it.

**Comment by JoanneDesign** [comment]
Luckily, so far, I have not seen it lock any other programs.

**Comment by AlexeiScherbakov** [comment]
I have similar problem.
1. It is not related only to dismount operations. It also can happened when user try to mount drive.
2. It more often happens with new veracrypt format than old truecrypt format.
3. It can happens with every app. VeraCrypt unfrezes when this app is closed or when VeraCrypt window is closed.

I have this bug with 1.19 & 1.21 versions of VeraCrypt

**Comment by idrassi** [comment]
On machines where such freeze occurs, checking the option "Don't show wait message dialog when performing operations" in "Settings -> Preferences" as mentioned by @JoanneDesign will fix the problem.

On my side, I can't reproduce the issue on different configuration. Maybe it is linked to common factor on those having the problem (for example, Antivirus software).

@CriusMacOldenly, @griesi007 : The option described above makes VeraCrypt doing all processing in the main thread without disabling any Windows. So, if the issue persists with this option checked, this means that something else is causing the freeze. Concerning the description you made about the issue affecting the second top most application, I will check the code related to this part and I will post my findings.

![VeraCrypt Settings-Preferences-Dont-show-wait-dialog-option](https://i.imgur.com/o0fzOEe.png)

**Comment by verybadsoldier** [comment]
@idrassi 
Enabling that option fixed the issue for me, thanks!

**Comment by leonard84** [comment]
The issue started after the applying the fall creators update, before that it ran fine.

@idrassi that workaround fixed the issue for me.

**Comment by CriusMacOldenly** [comment]
I have ensured this option is checked, and it does not fix the problem for me. Veracrypt stops all input from reaching the second top most application still. Mounting and dismounting does not restore input to the second top most window. The only way I have been able to recover input into the second top most application has been right clicking veracrypt, and telling it to exit. Once Veracrypt has exited, input can be done to the previously input disabled application again. Veracrypt still interferes with the second top most application by seemingly passing mouse clicks on itself to the second top most application. It's as if Veracrypt is treating itself, and the second top most application as the same window.

**Comment by tschombes** [comment]
I have the same issue with windows 10. Enabling the option described above does NOT help.
Veracrypt freezes the whole machine, i can do nothing.Sometimes it freezes the machine for 1 minute, sometimes i had to do a hard power off because it just freezes forever.
Very odd. With Truecrypt there was never a problem. I think i had to switch back to Truecrypt latest version :-(

**Comment by CriusMacOldenly** [comment]
It sounds like Windows itself is the second top most window for you tschombes. Are you trying to mount/dismount Veracrypt with the desktop only visible? It is very dangerous if you minimize all windows (by clicking the right most area of the taskbar for example) and then try to mount of dismount a drive in Veracrypt. I suggest you make sure you have an application window visible when mounting/dismounting a drive in Veracrypt to avoid stopping all input from reaching Windows 10 itself.

**Comment by CriusMacOldenly** [comment]
As a possible aid in debugging, I use Avast antivirus, and use an intel i7-6700k @4.00Ghz (not overclocked) and a gigabyte z170x ultra gaming motherboard. My video card is an nvidia geforce gtx 1070. I also am often running koplayer which is an android emulator. Are there other details that may help?

**Comment by tschombes** [comment]
CriusMacOldenly: is this a joke? "It is very dangerous if you minimize all windows when mounting/dismounting". What kind of shitty software is this? I'm going to delete it, because it is just not useable. TrueCrypt caused never andy problems!

**Comment by CriusMacOldenly** [comment]
It's not something I've proven yet, and it could be completely wrong. It's not a joke, but it is a warning. There is no doubt that Veracrypt is disabling input (via mouse at the very least) to windows that do not belong to Veracrypt. I have also proven (to myself at least) that telling Veracrypt to end restores input functionality to whatever windows were blocked from receiving input. Since the desktop itself is a window (as far as I know), it makes sense that Veracrypt would affect it the same way as any other window. You likely tried to use the mouse to interact with Windows. I don't remember if the keyboard is disabled or not.

**Comment by CriusMacOldenly** [comment]
tschombes - Truecrypt was great software, however, the software's authors have stated it is not secure, and they aren't going to bother making it secure. I seem to recall some security testers finding a hole in Truecrypt which Veracrypt fixes. Basically, you will be using unsupported software that is said to not do a good job by both the authors and independent testers. My recommendation is to just open a random program, like notepad, or calculator, and bring it to the foreground just before mounting or dismounting a drive. In this fashion, an application you don't care about will freeze rather than your window's window, or some application you care about. This work around works perfectly in all my tests.

**Comment by tschombes** [comment]
CriusMacOldenly - thanks for explanation, i will give it try. But to be honest - this freeze-issue exists since more than 1 year and there is no fix from the authors. this is just not understandable.

**Comment by CriusMacOldenly** [comment]
I know, I've had the problem for a very long time too.. The problem is, the Veracrypt authors can't reproduce the problem, and they've looked at the code trying to figure it out. By and large, most people's problems are fixed by telling Veracrypt to not show the wait message while performing operations. For you and I, this doesn't fix the problem.

As far as I know, the authors don't get paid to maintain this program -- the source code is freely available right here. You can compile it and test it yourself to see what's causing the problem.. I could too if I could read the code.. It's just volunteers who maintain the code now, right? I wouldn't put too much on their shoulders -- someone somewhere introduced this bug, and if the bug can't be seen by the coders, it can't be fixed.

**Comment by id10terr** [comment]
I have this problem too, on both my Win 10 machines.

It happens to me when mounting and unmounting consistently.
My workaround is to open a file explorer window and have it be my active window.
Then I do the veracrypt action, wait for the freeze, then use the taskbar right click 
on the file explorer icon to close the file explorer.
Veracrypt then comes back to life.

"Veracrypt stops all input from reaching the second top most application" is pretty much how I'd describe it. It is like the two two apps, VC and whatever was on top, are now linked in some input dependency and quitting the other unfreezes VC.

I've tried ticking the "Don't show wait ..." option.

**Comment by davebrny** [comment]
im having the same problem on windows 10. when dismounting it switches back to the last active window, then when i go back to veracrypt its frozen until i kill the last active window, then it dismounts properly 

other times i just click X to close the veracrypt window after a minute and windows forces it to shut down but im always worried the volume will become corrupt doing it this way

**Comment by CriusMacOldenly** [comment]
I have successfully frozen the desktop by mounting a drive in veracrypt. Usually I keep notepad open as the second top most window while mounting/dismounting veracrypt, and this was no exception. However, I had need to run an application that I had as a shortcut on my desktop. After veracrypt mounted the drive, I waited about a second, then minimized everything. I was about to start the application I needed via its desktop shortcut when suddenly the desktop flickered. Nothing I did on the desktop worked -- it was frozen. Windows itself continued to work fine, and I could use any other application. I needed the application I had on the desktop however, so I had to dismount and completely exit veracrypt to re-enable my desktop window again. Once I closed veracrypt, about 8-9 instances of the application suddenly opened up. It's obvious all my double clicks on the application were saved, and applied after veracrypt had exited.

In summary:
1. Veracrypt does not immediately freeze the windows. There is a delay measured in seconds before it freezes the window.
2. Everything you do to the frozen window is saved until veracrypt is closed. Then what you did to the window is applied.
3. Dismounting the drive, or mounting the drive will not unfreeze the window -- veracrypt has to actually end. If veracrypt ends, the frozen window is no longer frozen.
4. The frozen window is the topmost non veracrypt window, whatever it is. This can include the desktop, or other programs.

**Comment by phistrom** [comment]
@idrassi thank you so much for all your hard work on Veracrypt. Unfortunately, I still have this issue with 1.22-BETA5 installed. It behaves the same as previous versions for me.

My setup:
I have a 6TB hard drive. It has a GPT partition table. 5TB is allocated as a normal NTFS partition. The remaining 1TB is a Veracrypt AES-encrypted partition. It shows up as RAW in Windows Disk Manager. I have assigned it as a "favorite" volume.

1. Open Firefox - go to a website.
2. In Veracrypt, Mount favorite volumes.
3. Dismount favorite volumes in Veracrypt.

At this point both the Veracrypt and Firefox GUI is hung. A YouTube video in Firefox continues to play, but the X in the top right of the window as well as any buttons inside Firefox are now unclickable. If I kill the firefox.exe process, Veracrypt suddenly stops hanging and I get the notification that my volumes have been dismounted successfully. Firefox can now be opened and works just fine.

**Comment by idrassi** [comment]
@phistrom: Thank you for this detailed feedback. I think I have found
the real culprit behind this illusive issue: we used a non documented
way to implement setting VeraCrypt dialog as foreground Windows and it
appears that this approach is causing instability under Windows 10.

To avoid this, I have used an alternative approach that is less reliable
but less risky for stability
(https://github.com/veracrypt/VeraCrypt/commit/b2069885313881a177798375160256879c7b796c).

I have uploaded the installer for version 1.22-BETA6 that contains this
fix to
https://sourceforge.net/projects/veracrypt/files/VeraCrypt%20Nightly%20Builds/.

Can you please check that it fixes the issue on your side?
Thank you.

**Comment by phistrom** [comment]
@idrassi I am excited to report this issue is fixed for me! Hopefully others will confirm the same. This issue has been around for quite some time and I really cannot express enough my gratitude!

**Comment by AlexeiScherbakov** [comment]
Why not use 2 calls ofSetWindowPos:
First with HWND_TOPMOST and flags SWP_NOMOVE | SWP_NOSIZE
And second with HWND_NOTOPMOST and same flags? Look here:
https://stackoverflow.com/questions/916259/win32-bring-a-window-to-top

пн, 26 Мар 2018, 2:29 phistrom <notifications@github.com>:

**Comment by idrassi** [comment]
Thank you @phistrom for the confirmation. So far it looks good this time.

@AlexeiScherbakov: the approach described in your link uses |a
combination of AttachThreadInput,||SetForegroundWindow
and||SetWindowPos. The previous code in VeraCrypt was using a similar
approach (without ||SetWindowPos calls) and it looks that the stability
issues were caused by the use of ||AttachThreadInput on external
processes. That's why I completely changed the implementation and
resorted to less risky method.|||||**

**Comment by id10terr** [comment]
Hi,
I would also like to say this seems to have resolved the issue on my Win10 boxes.

Thanks!

**Comment by tsjensen** [comment]
I am using VeraCrypt **1.22** 64bit on Windows 10, and it still freezes the entire machine. The only way out is to power off the hard way, as all input is dead. Even the mouse cursor stops moving.
I am still in the process of trying out the various workarounds mentioned in this thread. Just thought I'd let you know that the issue is not resolved yet.

**Comment by CriusMacOldenly** [comment]
For me the issue got completely fixed. Did you look in the event viewer for application hang or similar messages?

**Comment by tsjensen** [comment]
Well, since the entire machine is completely frozen (including even mouse cursor movement), there is no looking at anything. Happens on mount as well as dismount, by the way.
But I'll check tomorrow if anything valuable was logged while the thing was dying. I use two machines, one of which is affected by the problem. On the other, all is well. Both are 1.22 64bit. So, maybe you're as lucky as my second machine.

**Comment by id10terr** [comment]
I've been using 1.23-Beta1 for a while now on two machine with no issues.

**Comment by tsjensen** [comment]
@idrassi Thank you for the quick response!
You are correct, it happens on both mount and dismount, but not every time. That why I need some more time to try the workarounds, to be sure they actually fix the problem. Will report back here.
The affected machine is the work machine, which has all kinds of special stuff on it I can't influence, as is usual for big companies. No Acronis though, and also no BSOD symptom, just freezing.

**Comment by tsjensen** [comment]
Just a quick follow-up: The problem is solved for me as I have changed the way I call VeraCrypt.  
I now open a dedicated cmd window, and call a batch file from there, which in turn calls VeraCrypt. I still see temporary freezing during mount and/or dismount operations on random applications which are open at the same time, but it doesn't freeze the entire system anymore. Still, this behavior feels buggy.

**Comment by junpeikyu** [comment]
Happend to me too. It freezes the "active and not minimized" Program-Window. Solution for me was: only have Veracrypt maximized and dismount.

**Comment by rupertbayern** [comment]
Similar problem here. I updated to 1.23 and I still have this problem.

1) Boot PC with Windows 10 (veracrypt is in autostart)
2) Mount favourite volumes
3) Dismount one volume
4) Whole GUI turns white and stops responding.

The taskmanager sayst that VeraCrypt is "inactive" and it barely uses any system resources. The volume doesn't get dismounted, I can still see the drive in the Explorer and I can still access all data.

I set VeraCrypt to not show waiting dialog and I didnt start any other program and this problem still persits. As long as I dont restart the PC I the drive stays mounted. This happens every time.

**Comment by RnoldR** [comment]
I started today using veracrypt on Ubuntu 18.04 and windows 10. On both systems it freezes, especially when I start to backup to the container. As the container is quite big (2TB) it takes ours to format it. When after that it freezes during backup and the data is lost as well I took the decision to switch back to truecrypt. Veracrypt has some advantages such as creating ntfs formatted containers under linux and being (allegedly) more secure, while truecrypt sometimes has troubles to elevate to administrator level, but this freezing business is very frustrating. I am willing to test though for ubuntu, just let me know.

I noticed that the preferences window is different in Ubuntu, I could not ind the option not to show the wait message and some other options. Furthermore I could not tab from one field to another and last but not least, I couldn't correct the text I typed in the password field. 

When Veracrypt and applications I had started before starting veracrypt froze, I logged off and on again. When I started Keepass it froze as well. I did a `ps` and found the following processes running:

```
$ ps -ef | grep veracrypt
arnold     2344      1  0 19:59 ?        00:00:00 veracrypt
root       2645      1  0 19:59 ?        00:00:01 /usr/bin/veracrypt --core-service
root       2648      1  0 19:59 ?        00:00:00 /usr/bin/veracrypt --core-service
root       2650      1  0 19:59 ?        00:00:03 /usr/bin/veracrypt --core-service
root       2738      1 12 19:59 ?        00:06:07 /sbin/mount.ntfs /dev/mapper/veracrypt1 /media/veracrypt1 -o rw,uid=1000,gid=1000,umask=077
root       8458   2645  0 20:29 ?        00:00:00 umount -- /media/veracrypt1
arnold    11093      1  0 20:42 ?        00:00:01 veracrypt
arnold    11094  11093  0 20:42 ?        00:00:00 veracrypt
arnold    12805  12782  0 20:47 pts/0    00:00:00 grep --color=auto veracrypt
```

Hope this helps.

**Comment by alt3r-3go** [comment]
Is this still a problem for anyone on either the latest released 1.23-Hotfix2 or the latest pre-released 1.24-Beta5? If yes, please describe your configuration (OS version, VC version, and outline of what "special" software is installed like antiviruses or other sorts of endpoint protection, DRM/ERM or somesuch), so that I could try to reproduce it.

**Comment by CriusMacOldenly** [comment]
This problem was fixed for me on or near Aug 27, 2018. The problem has not returned.

**Comment by GH-Rake** [comment]
@CriusMacOldenly I recommend upgrading, the latest version came out in September and it's awesome

**Comment by CriusMacOldenly** [comment]
But...but... the Aug 27, 2018 version works perfectly... :(

**Comment by alt3r-3go** [comment]
@CriusMacOldenly thanks, that's interesting. Maybe that was (at least in your case) something else than VC at the end of the day like an antivirus, that got updated, or something.

Let's see if anyone else still sees this.

**Comment by CriusMacOldenly** [comment]
@alt3r-3go No, it was 100% the update by idrassi on Mar 25, 2018 that completely fixed the issue.
I had the problem just before I downloaded that beta, and I never saw the problem after that update.

**Comment by alt3r-3go** [comment]
Oh, I see, thanks for clarification. I understood it in a way that it self-fixed after that August 27 without any changes at the VC side, but you meant that you have none of the two versions I've suggested to try as the earlier fix corrected it for you. Got it.

**Comment by id10terr** [comment]
1.22-BETA6 fixed the issue for me, but I've been using 1.23 Hotfix2 since last year and had no problems.
So much so, I haven't even checked for later versions.

thanks!

**Comment by alt3r-3go** [comment]
Wonderful, thanks @id10terr. @idrassi, looks like this one could be closed then as the latest official release doesn't have this problem (as we see it was fixed even before that in one of the betas).

**Comment by VernonStow** [comment]
I am having this problem with VC 1.24 Update 2 on Windows 10 Home. When I run VC and back up two separate encrypted USB drives, it usually hangs when I try to dismount the _second_ drive and also causes explorer.exe to hang.

I have to kill both processes with Task Manager, then restart explorer.exe and VC, which still shows the second drive mounted. At this point, I can unmount the USB drive normally. No errors are logged by Windows. I still have this problem after clean reinstalling VC.

**Comment by alt3r-3go** [comment]
@VernonStow, what kind of antivirus do you have? You're saying "usually" - so it's not hanging all the time? Could you please provide a ballpark estimate of how many times out of ten it typically hangs?

Finally, could you please check if there's any process holding files on the mounted container, before you try to unmount it? You can e.g. use [Sysinternals Process Explorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) for that (Handle menu item).

**Comment by VernonStow** [comment]
> 
> 
> @VernonStow, what kind of antivirus do you have? You're saying "usually" - so it's not hanging all the time? Could you please provide a ballpark estimate of how many times out of ten it typically hangs?
> 
> Finally, could you please check if there's any process holding files on the mounted container, before you try to unmount it? You can e.g. use [Sysinternals Process Explorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) for that (Handle menu item).

I use Windows Defender, no other security software. The hang on dismount happens about 2/3 of the time. I noticed Windows Explorer ("This PC") was unresponsive the previous time VC hung, but today I had the same dismounting problem with VC even though I used a different drive letter for the second USB container and closed the Explorer window first. I will check it in Process Explorer the next time it happens.

**Comment by VernonStow** [comment]
It happened again today. I closed VC completely after dismounting the first USB drive, then reopened it, did my second backup, and VC froze on dismount. I opened up Process Explorer and made a minidump file from the VC process (taking a total of about 2 minutes), when suddenly VC unfroze and dismounted.

I have the .dmp file, but couldn't get it to open with Nirsoft's Blue Screen View. Would this file contain confidential information like filenames, or could I safely upload it here?

**Comment by crius-al** [comment]
> 
> 
> It happened again today. I closed VC completely after dismounting the first USB drive with no problem. I restarted VC, did my second backup, and VC froze on dismount. I opened Process Explorer and made a minidump file from the VC process (it had been a total of about 2 minutes by this point), when suddenly VC unfroze and dismounted.
> 
> I have the .dmp file, but couldn't get it to open with Nirsoft's Blue Screen View. Would this file contain confidential information like filenames, or could I safely upload it here?

Nirsoft's Blue Screen Viewer is for viewing Blue Screen of Death files, which didn't happen in your case.
What you want is windbg.exe. This is built into windows for looking at .dmp files.

**Comment by CriusMacOldenly** [comment]
> 
> 
> It happened again today. I closed VC completely after dismounting the first USB drive with no problem. I restarted VC, did my second backup, and VC froze on dismount. I opened Process Explorer and made a minidump file from the VC process (it had been a total of about 2 minutes by this point), when suddenly VC unfroze and dismounted.
> 
> I have the .dmp file, but couldn't get it to open with Nirsoft's Blue Screen View. Would this file contain confidential information like filenames, or could I safely upload it here?

Nirsoft's Blue Screen Viewer is for viewing Blue Screen of Death files, which didn't happen in your case.
What you want is windbg.exe. This is built into windows for looking at .dmp files.

**Comment by alt3r-3go** [comment]
I wouldn't share minidump, not right away at least. Let's just get the used files first (search by the "<your disk letter>:" there), that should provide the additional information I'm looking for without revealing any sensitive info.

**Comment by VernonStow** [comment]
I could not find that information with windbg.exe (downloaded from the MS store). Could you tell me specifically where to look.

(Just to note, the next time I mounted/dismounted the same USB drives to run backups, there was no crash or hang.)

**Comment by alt3r-3go** [comment]
Windbg won't show you that, try the one I suggested in one of my previous comments, [Sysinternals Process Explorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer), see Handle menu item.

**Comment by VernonStow** [comment]
I have not been able to find a display of handles using Process Explorer, just a count of how many. The only Microsoft documentation seems to be to purchase a book.

I have observed the hanging problem with two different Kingston SE9 USB drives but not with any other brand at this point. If I leave it undisturbed for 5-10 minutes, it will eventually pop up a message that the drive is in use and then dismount.

**Comment by GH-Rake** [comment]
I believe ProcessHacker will list all the handles

**Comment by VernonStow** [comment]
> 
> 
> I believe ProcessHacker will list all the handles

Thanks, I've downloaded a copy and will use it when needed. Curiously, the last two times I dismounted the SE9 USB drive from VC, I did not experience any freeze or hang.

**Comment by VernonStow** [comment]
> 
> 
> I wouldn't share minidump, not right away at least. Let's just get the used files first (search by the `<your disk letter>:` there), that should provide the additional information I'm looking for without revealing any sensitive info.

[Handles from VC hang.txt](https://github.com/veracrypt/VeraCrypt/files/4285046/Handles.from.VC.hang.txt)

I used Process Hacker to obtain the handles in use during a VC hang, copy/pasted into the text file attached.

**Comment by alt3r-3go** [comment]
@VernonStow sorry, but that is not helpful - it looks like the runtime object dump for the VC process, this is not what I'm talking about. If you're using ProcessHacker, then Ctrl+F would bring up the window for finding processes holding files on your mounted drive - see the "Can't edit or delete a file <...>" section on the ProcessHacker's [home page](https://processhacker.sourceforge.io/). That's the same functionality as Process Explorer provides and that's what I mean.

**Comment by VernonStow** [comment]
@alt3r-3go OK, will do.

**Comment by alt3r-3go** [comment]
@VernonStow, I've found a video on YouTube that demonstrates the functionality I'm talking about (and by Mark Russinovich himself, no less :smiley:). See here: https://youtu.be/YGtsMa9wbjw?t=6079

**Comment by VernonStow** [comment]
alt3r-3go, Thanks! That was extremely helpful.

I captured two text files with Process Explorer showing the handles before and after attempting to dismount. After a 5-minute hang, it eventually dismounted without an error message.

[VeraCrypt-x64.exe before dismount.txt](https://github.com/veracrypt/VeraCrypt/files/4335845/VeraCrypt-x64.exe.before.dismount.txt)
[VeraCrypt-x64.exe during hang after dismount.txt](https://github.com/veracrypt/VeraCrypt/files/4335846/VeraCrypt-x64.exe.during.hang.after.dismount.txt)

**Comment by alt3r-3go** [comment]
Sorry, that's still not the right one. You've dumped the VeraCrypt's objects while I'm talking about the opposite thing - checking what *process* holds locks on your drive mounted by VeraCrypt. 

You'd need to do something like:
1. Open ProcessExplorer, so that it's ready to work when needed
1. Start unmount process in VC
1. If this hangs - switch to ProcessExplorer, click the Find menu item and then in the "Handle or DLL substring" edit box type your VC drive letter and a colon, e.g. "D:", then click Search.
1. That should provide you with a list of all processes working with your D: drive. E.g. on my example screenshot below that's Notepad having a file open. The Find menu is also visible on the screenshot.
1. Make a screenshot or just record what's there - my guess is that there's some application that's holding those locks and preventing VC from properly unmounting. I've seen things like that in my daily use of VC, but it has nothing to do with it.

![image](https://user-images.githubusercontent.com/45139988/77826981-4ab64d00-7113-11ea-8441-33683b6966ee.png)

**Comment by VernonStow** [comment]
Update: Another laptop user in a personal conversation mentioned the same issue to me, that is, VeraCrypt portable freezing on trying to dismount a Kingston brand USB drive. The user said the freeze did not occur on a desktop PC, only on a laptop and only with a Kingston drive. I have had this issue with several different Kingston SE9 USB drives on a laptop (my only computer), but not with any other brand of USB drive. I suspect that the controller on the USB drives of this brand may be at fault.

**Comment by Giger22** [comment]
I have still the same problem. Artix linux 6.1.8

**Comment by clst** [comment]
I just had the same issue. After looking everywhere for the reason I noticed an explorer.exe process (even though no explorer windows were open).

Commandline: `C:\Windows\explorer.exe /factory,{ceff45ee-someGUID-a022c81eda92} -Embedding`

After killing it, the volume immediately dismounted.

**Comment by riorao** [comment]
проблема актуальна на портативной версии 1.26.20 и win11  24H2
при закрытии приложения не даёт извлечь накопитель из за того что veracrypt-x64.sys находится в памяти.
